File: frontend/lib/models/adopter_form_model.dart
class AdopterFormModel {
  String housingType;
  String housingStatus;
  bool hasYard;
  bool hasProtectiveNetting;
  String householdMembers;
  bool allMembersAgree;
  bool hasAllergies;
  int hoursPetAlonePerDay;
  bool hasOtherPets;
  String? otherPetsDetails;
  String? previousPetExperience;
  bool monthlyBudgetConfirmed;
  bool emergencyFundAvailable;
  bool agreesToFollowUp;
  bool agreesToMandatoryNeutering;
  String? evaluationStatus;
  double? evaluationScore;

  AdopterFormModel({
    this.housingType = 'apartment',
    this.housingStatus = 'owned',
    this.hasYard = false,
    this.hasProtectiveNetting = true,
    this.householdMembers = 'alone',
    this.allMembersAgree = true,
    this.hasAllergies = false,
    this.hoursPetAlonePerDay = 4,
    this.hasOtherPets = false,
    this.otherPetsDetails,
    this.previousPetExperience,
    this.monthlyBudgetConfirmed = true,
    this.emergencyFundAvailable = true,
    this.agreesToFollowUp = true,
    this.agreesToMandatoryNeutering = true,
    this.evaluationStatus,
    this.evaluationScore,
  });

  factory AdopterFormModel.fromJson(Map<String, dynamic> json) {
    return AdopterFormModel(
      housingType: json['housing_type'] as String? ?? 'apartment',
      housingStatus: json['housing_status'] as String? ?? 'owned',
      hasYard: json['has_yard'] as bool? ?? false,
      hasProtectiveNetting: json['has_protective_netting'] as bool? ?? false,
      householdMembers: json['household_members'] as String? ?? 'alone',
      allMembersAgree: json['all_members_agree'] as bool? ?? true,
      hasAllergies: json['has_allergies'] as bool? ?? false,
      hoursPetAlonePerDay: json['hours_pet_alone_per_day'] as int? ?? 4,
      hasOtherPets: json['has_other_pets'] as bool? ?? false,
      otherPetsDetails: json['other_pets_details'] as String?,
      previousPetExperience: json['previous_pet_experience'] as String?,
      monthlyBudgetConfirmed: json['monthly_budget_confirmed'] as bool? ?? true,
      emergencyFundAvailable: json['emergency_fund_available'] as bool? ?? true,
      agreesToFollowUp: json['agrees_to_follow_up'] as bool? ?? true,
      agreesToMandatoryNeutering:
          json['agrees_to_mandatory_neutering'] as bool? ?? true,
      evaluationStatus: json['evaluation_status'] as String?,
      evaluationScore: (json['evaluation_score'] is num)
          ? (json['evaluation_score'] as num).toDouble()
          : null,
    );
  }

  String get housingTypeLabel {
    switch (housingType) {
      case 'house':
        return 'Casa';
      case 'apartment':
        return 'Departamento';
      case 'farm':
        return 'Chacra / Campo';
      default:
        return 'Otro';
    }
  }

  String get housingStatusLabel {
    switch (housingStatus) {
      case 'owned':
        return 'Propietario/a';
      case 'rented_allowed':
        return 'Alquiler (mascotas permitidas)';
      case 'rented_pending_permission':
        return 'Alquiler (permiso pendiente)';
      default:
        return housingStatus;
    }
  }

  String get householdMembersLabel {
    switch (householdMembers) {
      case 'alone':
        return 'Vive solo/a';
      case 'couple':
        return 'En pareja';
      case 'family_with_young_kids':
        return 'Familia con niños pequeños';
      case 'family_with_teens':
        return 'Familia con adolescentes';
      case 'roommates':
        return 'Con compañeros de casa';
      default:
        return householdMembers;
    }
  }

  Map<String, dynamic> toJson() {
    return {
      'housing_type': housingType,
      'housing_status': housingStatus,
      'has_yard': hasYard,
      'has_protective_netting': hasProtectiveNetting,
      'household_members': householdMembers,
      'all_members_agree': allMembersAgree,
      'has_allergies': hasAllergies,
      'hours_pet_alone_per_day': hoursPetAlonePerDay,
      'has_other_pets': hasOtherPets,
      'other_pets_details': otherPetsDetails,
      'previous_pet_experience': previousPetExperience,
      'monthly_budget_confirmed': monthlyBudgetConfirmed,
      'emergency_fund_available': emergencyFundAvailable,
      'agrees_to_follow_up': agreesToFollowUp,
      'agrees_to_mandatory_neutering': agreesToMandatoryNeutering,
    };
  }
}

File: backend/src/models/adopter_form.model.ts
import { z } from 'zod';
import { supabaseAdmin } from '../config/supabase.js';

// Esquema de validación estricto para el formulario de adopción
export const AdopterFormSchema = z.object({
  housing_type: z.enum(['house', 'apartment', 'farm', 'other']),
  housing_status: z.enum(['owned', 'rented_allowed', 'rented_pending_permission']),
  has_yard: z.boolean(),
  has_protective_netting: z.boolean(),
  household_members: z.enum(['alone', 'couple', 'family_with_young_kids', 'family_with_teens', 'roommates']),
  all_members_agree: z.boolean(),
  has_allergies: z.boolean(),
  hours_pet_alone_per_day: z.number().min(0).max(24),
  has_other_pets: z.boolean(),
  other_pets_details: z.string().optional().nullable(),
  previous_pet_experience: z.string().optional().nullable(),
  monthly_budget_confirmed: z.boolean(),
  emergency_fund_available: z.boolean(),
  agrees_to_follow_up: z.boolean(),
  agrees_to_mandatory_neutering: z.boolean(),
});

export type AdopterFormInput = z.infer<typeof AdopterFormSchema>;

export interface AdopterFormRecord extends AdopterFormInput {
  id: string;
  user_id: string;
  evaluation_status: 'pending' | 'approved' | 'requires_interview' | 'rejected';
  evaluation_score: number;
  reviewer_notes?: string | null;
  created_at: string;
  updated_at: string;
}

export class AdopterFormModel {
  /**
   * Algoritmo de idoneidad y scoring del adoptante:
   * Evalúa factores de riesgo (p.ej. horas solo, permiso de alquiler, redes en balcón)
   * y otorga un puntaje sobre 100.
   */
  static calculateSuitabilityScore(data: AdopterFormInput): { score: number; status: 'approved' | 'requires_interview' | 'rejected' } {
    let score = 50;

    // 1. Vivienda
    if (data.housing_status === 'rented_pending_permission') {
      score -= 25; // Riesgo de reclamo de desalojo
    } else {
      score += 10;
    }

    if (data.has_yard) score += 10;
    if (data.housing_type === 'apartment' && data.has_protective_netting) score += 10;

    // 2. Convivencia
    if (!data.all_members_agree) {
      return { score: 10, status: 'rejected' }; // Condición excluyente: todo el hogar debe estar de acuerdo
    }

    if (data.has_allergies) {
      score -= 10; // Requiere verificación o prueba de convivencia previa
    }

    // 3. Tiempo de soledad
    if (data.hours_pet_alone_per_day > 10) {
      score -= 20;
    } else if (data.hours_pet_alone_per_day <= 5) {
      score += 10;
    }

    // 4. Compromiso económico y veterinario
    if (!data.monthly_budget_confirmed || !data.emergency_fund_available) {
      score -= 25;
    } else {
      score += 10;
    }

    if (!data.agrees_to_follow_up || !data.agrees_to_mandatory_neutering) {
      return { score: 15, status: 'rejected' };
    }

    // Limitar entre 0 y 100
    score = Math.max(0, Math.min(100, score));

    let status: 'approved' | 'requires_interview' | 'rejected' = 'approved';
    if (score < 40) {
      status = 'rejected';
    } else if (score < 70) {
      status = 'requires_interview';
    }

    return { score, status };
  }

  static async findByUserId(userId: string): Promise<AdopterFormRecord | null> {
    const { data, error } = await supabaseAdmin
      .from('adopter_forms')
      .select('*')
      .eq('user_id', userId)
      .maybeSingle();

    if (error) throw error;
    return data as AdopterFormRecord | null;
  }

  static async upsertForm(userId: string, input: AdopterFormInput): Promise<AdopterFormRecord> {
    const evaluation = this.calculateSuitabilityScore(input);

    const payload = {
      user_id: userId,
      ...input,
      evaluation_score: evaluation.score,
      evaluation_status: evaluation.status,
      updated_at: new Date().toISOString(),
    };

    const { data, error } = await supabaseAdmin
      .from('adopter_forms')
      .upsert(payload, { onConflict: 'user_id' })
      .select()
      .single();

    if (error) throw error;
    return data as AdopterFormRecord;
  }
}
